File: Backend/HealthCareSystem/src/main/java/com/healthcare/service/AdminService.java
package com.healthcare.service;

import com.healthcare.dto.Appointment;
import com.healthcare.dto.DiagnosticCenter;
import com.healthcare.dto.Test;
import com.healthcare.dto.User;
import com.healthcare.repository.AppointmentRepository;
import com.healthcare.repository.DiagnosticCenterRepository;
import com.healthcare.repository.TestRepository;
import com.healthcare.repository.UserRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Objects;

@Service
public class AdminService {
    @Autowired
    private DiagnosticCenterRepository diagnosticCenterRepository;

    @Autowired
    private TestRepository testRepository;

    @Autowired
    private AppointmentRepository appointmentRepository;

    @Autowired
    private UserRepository userRepository;

    public User register(User user) {
        return userRepository.save(user);
    }

    public List<User> getAllUsers(){
        return userRepository.findAll();
    }

    public DiagnosticCenter addCenter(DiagnosticCenter center) {
        return diagnosticCenterRepository.save(center);
    }

    public List<DiagnosticCenter> getAllCenters(){
        return diagnosticCenterRepository.findAll();
    }

    public void removeCenter(int centerId) {
        diagnosticCenterRepository.deleteById(centerId);
    }

//    public Test addTest(Test test) {
//        return testRepository.save(test);
//    }

    public Test addTest(Integer centerId, String testName) {
        DiagnosticCenter diagnosticCenter = diagnosticCenterRepository.findById(centerId).orElseThrow();
        Test test = new Test();
        test.setTestName(testName);
        test.setDiagnosticCenter(diagnosticCenter);
        return testRepository.save(test);
    }

    public User login(String email,String password){
        List<User> users = getAllUsers();
        return users.stream().filter(o -> {
            return (o.getEmail().equals(email))
                    && o.getPassword().equals(password);
        }).findFirst().get();
    }

    public void removeTest(Long testId) {
        testRepository.deleteById(testId);
    }

    public List<Appointment> getAppointments(int centerId) {
        return appointmentRepository.findByDiagnosticCenterCenterId(centerId);
    }

    public void approveAppointment(int appointmentId, boolean approved) {
        Appointment appointment = appointmentRepository.findById(appointmentId).orElseThrow();
        appointment.setApproved(approved);
        appointmentRepository.save(appointment);
    }
    // Add a new method for choosing a center
    public DiagnosticCenter chooseCenter(int centerId) {
        return diagnosticCenterRepository.findById(centerId).orElseThrow();
    }

    // Add a new method for entering test details
    public void enterTestDetails(int centerId, Test test) {
        DiagnosticCenter center = diagnosticCenterRepository.findById(centerId).orElseThrow();
    }

    // Add a new method for submitting test details
//    public void submitTestDetails(int centerId) {
//        // Get the diagnostic center
//        DiagnosticCenter center = diagnosticCenterRepository.findById(centerId).orElseThrow();
//
//        // Get all tests associated with the center
//        //List<Test> tests = testRepository.findByDiagnosticCenterCenterId(centerId);
//
//        // Save each test to the database
//        for (Test test : tests) {
//            testRepository.save(test);
//        }
//    }


    // Get test details
    public List<Test> getAllTests() {
        return testRepository.findAll();
    }


    // Removes a test from a specific diagnostic center by test ID and center ID
//    public void removeTest(int centerId, Long testId) {
//        DiagnosticCenter center = diagnosticCenterRepository.findById(centerId)
//                .orElseThrow(() -> new IllegalArgumentException("Invalid center ID"));
//        Test test = testRepository.findById(testId)
//                .orElseThrow(() -> new IllegalArgumentException("Invalid test ID"));
//        if (test.getDiagnosticCenter().equals(center)) {
//            testRepository.delete(test);
//        } else {
//            throw new IllegalArgumentException("Test does not belong to the given center");
//        }
//    }

}

File: healthcare-system-frontend/src/app/service/admin.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { User } from '../models/user.model';
import { DiagnosticCenter } from '../models/diagnostic-center.model';
import { Test } from '../models/test.model';
import { Appointment } from '../models/appointment.model';

@Injectable({
  providedIn: 'root'
})
export class AdminService {
  private apiUrl = 'http://localhost:8080/admin';

  constructor(private http: HttpClient) { }

  register(user: User): Observable<User> {
    return this.http.post<User>(`${this.apiUrl}/register`, user);
  }

  getAllUsers(): Observable<User[]> {
    return this.http.get<User[]>(`${this.apiUrl}/allUsers`);
  }

  addCenter(center: DiagnosticCenter): Observable<DiagnosticCenter> {
    return this.http.post<DiagnosticCenter>(`${this.apiUrl}/addCenter`, center);
  }

  getAllCenters(): Observable<DiagnosticCenter[]> {
    return this.http.get<DiagnosticCenter[]>(`${this.apiUrl}/allCenters`);
  }

  removeCenter(centerId: string): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/centers/${centerId}`);
  }

  addTest(test: Test): Observable<Test> {
    return this.http.post<Test>(`${this.apiUrl}/addTest`, test);
  }

  removeTest(testId: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/tests/${testId}`);
  }

  getAppointments(centerId: string): Observable<Appointment[]> {
    return this.http.get<Appointment[]>(`${this.apiUrl}/centers/${centerId}/appointments`);
  }

  approveAppointment(appointmentId: number, approved: boolean): Observable<void> {
    return this.http.post<void>(`${this.apiUrl}/appointments/${appointmentId}/approve`, { approved });
  }
}
